src/nupic/algorithms/Cells4.cpp

```diff
@@ -1628,46 +1667,55 @@ void Cells4::adaptSegment(const SegmentUpdate& update)
     // updateSynapses? It would be cleaner and avoid the need to keep the
     // synapses sorted.
 
-    // Accumulate list of synapses to decrement, increment, add, and remove
-    std::set<UInt> synapsesSet(update.begin(), update.end());
+    // Accumulate lists of synapses to decrement, increment, add, and remove
+    std::set<UInt> synapsesSet(update.begin(), update.end()); // srcCellIdx
     static std::vector<UInt> removed; // srcCellIdx
-    static std::vector<UInt> synToDec, synToInc,
-                             inactiveSegmentIndices, activeSegmentIndices;
-    removed.clear() ;                       // purge residual data
-    synToDec.clear() ;                      // purge residual data
-    synToInc.clear() ;                      // purge residual data
-    inactiveSegmentIndices.clear() ;        // purge residual data
-    activeSegmentIndices.clear() ;          // purge residual data
-    for (UInt i = 0; i != segment.size(); ++i) {
-      UInt srcCellIdx = segment[i].srcCellIdx();
-      if (not_in(srcCellIdx, synapsesSet)) {
-        synToDec.push_back(srcCellIdx);  // TODO: Check synapse still exists!
-        inactiveSegmentIndices.push_back(i);
-      }
-      else {
-        synToInc.push_back(srcCellIdx);
-        synapsesSet.erase(srcCellIdx);
-        activeSegmentIndices.push_back(i);
-      }
-    }
+    static std::vector<UInt> synToDec, synToInc; // srcCellIdx
+    static std::vector<UInt> inactiveSegmentIndices, activeSegmentIndices; // syn idx
+    removed.clear();                        // purge residual data
+
+    _generateListsOfSynapsesToAdjustForAdaptSegment(
+      segment, synapsesSet,
+      synToDec, inactiveSegmentIndices,
+      synToInc, activeSegmentIndices);
 
-    // Now update synapses which need to be decremented or incremented
+    // Decrement permanences of inactive synapses
     // TODO: Why can't we just do this inline in the above loop?
```

Comment:
@scottpurdy, I am going to remove this TODO for a couple of reasons:

1. Based on my current understanding, however limited, it may not be practical to do what's being suggested in the TODO from performance perspective, or it may require violating data encapsulation.
2. As you have explained previously, this code is in bug-fix mode, so making this sort of change would be inconsistent with our priorities.